### DISCUSSION TITLE
Freeze models

### DISCUSSION THREAD

**Discussion by davfsa** [hikari-py/hikari#1110] [Ideas]
This idea was originally proposed by @FasterSpeeding in https://github.com/hikari-py/hikari/pull/611, but it got closed in the end due to a performance difference between frozen and un-frozen attr classes (about 300ns on a laptop according to the documentation).

I recently wanted to see if I could get that difference down and ended up speeding creation of frozen classes by 10%, closing the gap a bit more (https://github.com/python-attrs/attrs/pull/898). Might try and reduce it a bit more in the future, but I highly doubt I will get anywhere.

*Note that there is no release on PyPI with that code just yet*

Even with that performance hit, we could freeze all models to ensure inmutability and make the cache more efficient by removing all the copy calls, which https://github.com/hikari-py/hikari/pull/611 did.

**Comment by FasterSpeeding** [comment]
Under the current design there's one outlier of `Embed` which can't be frozen (as of present) due to being both a builder and a data class.

**Comment by rayakame** [comment]
this still needs to be implemented, but probably in the future together with the msgspec rewrite
